### PR TITLE
Check list of available topics before publishing a message

### PR DESCRIPTION
**Pull request by agrare**
To prevent messages destined for non-existant topics blocking other valid messages from being delivered, check the list of topics (cached) before publishing a message.

### CODE REVIEW

File: lib/sources/api/messaging.rb [hunk 1]
```diff
@@ -1,16 +1,24 @@
...
         require "manageiq-messaging"
 
-        @client ||= {}
-        @client[topic] ||= ManageIQ::Messaging::Client.open(
-          :protocol => :Kafka,
-          :host     => ENV["QUEUE_HOST"] || "localhost",
-          :port     => ENV["QUEUE_PORT"] || "9092",
-          :encoding => "json"
-        )
+        Thread.current[:messaging_client] ||= begin
```
Comment (agrare) [inline lib/sources/api/messaging.rb]: Went back to caching per thread but want to leave this at 1 puma thread first to test one variable at a time

**Comment by carbonin** [comment]
So this is now a single client per thread (of which we only have one) as opposed to one per topic, right?
But in this case we're ensuring that we don't try to publish to topics that don't exist?

File: app/controllers/api/v1/sources_controller.rb [hunk 2]
```diff
@@ -23,19 +23,24 @@ def check_availability
...
+              }
             }
-          }
-        )
-        logger.debug("publishing message to #{topic}...Complete")
+          )
+
+          logger.debug("publishing message to #{topic}...Complete")
+        else
+          logger.debug("not publishing message to non-existing topic: #{topic}")
```
Comment (carbonin) [inline app/controllers/api/v1/sources_controller.rb]: Ideally this shouldn't happen ... can we make this an error or warn?

**Comment by agrare** [comment]
> So this is now a single client per thread (of which we only have one) as opposed to one per topic, right?
> But in this case we're ensuring that we don't try to publish to topics that don't exist?

@carbonin yes exactly

File: app/controllers/api/v1/sources_controller.rb (hunk 2 above)
Comment (agrare) [reply app/controllers/api/v1/sources_controller.rb]: So the way the availability checker works right now this actually happens rather often, and not just for source-types that we don't have an operations worker for yet.

E.g. There are the cloudforms source-types like ovirt, vsphere, openstack where I don't think we'll ever have an operations worker.

Comment (agrare) [reply app/controllers/api/v1/sources_controller.rb]: I think the way to properly fix this, is to turn this current check into the topological-inventory application check.  So sources that aren't enabled for topology don't try to write to this topic.  Then topology is playing at the same level as all the other apps on the platform.

Comment (carbonin) [reply app/controllers/api/v1/sources_controller.rb]: Yeah, I was thinking this was happening pretty often, but we log debug by default right now anyway so I was thinking making this an error would make the failing topics easier to find.

Comment (agrare) [reply app/controllers/api/v1/sources_controller.rb]: :+1: done

**Pull request merged**